Missing error checks for fread

```
In file included from /home/marcin/.cache/pacaur/ivan/src/ivan/Main/Include/bodypart.h:17:0,
                 from /home/marcin/.cache/pacaur/ivan/src/ivan/Main/Include/char.h:16,
                 from /home/marcin/.cache/pacaur/ivan/src/ivan/Main/Source/dataset.cpp:13:
/home/marcin/.cache/pacaur/ivan/src/ivan/FeLib/Include/save.h: In member function ‘void inputfile::Read(char*, long int)’:
/home/marcin/.cache/pacaur/ivan/src/ivan/FeLib/Include/save.h:75:70: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
   void Read(char* Offset, long Size) { fread(Offset, 1, Size, Buffer); }
```
Since it's C++, you may want to use `fstream`. 
If you assume no error should appear, it's better to crash the whole thing to hell than to experience strange bugs.

--- a/FeLib/Include/save.h
+++ b/FeLib/Include/save.h
@@ -14,7 +14,7 @@
 #define __SAVE_H__
 
 #include <ctime>
-#include <cstdio>
+#include <fstream>
 #include <vector>
 #include <deque>
 #include <list>
@@ -40,22 +40,16 @@ inline inputfile& operator>>(inputfile& SaveFile, type& Value)\
 
 typedef std::map<festring, long> valuemap;
 
-/* fstream seems to bug with DJGPP, so we use FILE* here */
-
 class outputfile
 {
  public:
   outputfile(cfestring&, truth = true);
-  ~outputfile();
-  void Put(char What) { fputc(What, Buffer); }
-  void Write(cchar* Offset, long Size)
-  { fwrite(Offset, 1, Size, Buffer); }
-  truth IsOpen() { return Buffer != 0; }
-  void Close() { fclose(Buffer);  Buffer = 0; }
-  void Flush() { fflush(Buffer); }
-  void ReOpen();
+  void Put(char What) { File.put(What); }
+  void Write(cchar* Data, long Size) { File.write(Data, Size); }
+  truth IsOpen() { return File.is_open(); }
+  void Close() { File.close(); }
  private:
-  FILE* Buffer;
+  std::ofstream File;
   festring FileName;
 };
 
@@ -63,7 +57,6 @@ class inputfile
 {
  public:
   inputfile(cfestring&, const valuemap* = 0, truth = true);
-  ~inputfile();
   festring ReadWord(truth = true);
   void ReadWord(festring&, truth = true);
   char ReadLetter(truth = true);
@@ -71,23 +64,22 @@ class inputfile
   festring ReadStringOrNumber (long *num, truth *isString, truth PreserveTerminator=false);
   v2 ReadVector2d();
   rect ReadRect();
-  int Get() { return fgetc(Buffer); }
-  void Read(char* Offset, long Size) { fread(Offset, 1, Size, Buffer); }
-  truth IsOpen() { return Buffer != 0; }
-  truth Eof() { return feof(Buffer); }
-  void ClearFlags() { clearerr(Buffer); }
-  void SeekPosBegin(long Offset) { fseek(Buffer, Offset, SEEK_SET); }
-  void SeekPosCurrent(long Offset) { fseek(Buffer, Offset, SEEK_CUR); }
-  void SeekPosEnd(long Offset) { fseek(Buffer, Offset, SEEK_END); }
-  long TellPos() { return ftell(Buffer); }
+  int Get() { return File.get(); }
+  void Read(char* Data, long Size) { File.read(Data, Size); }
+  truth IsOpen() { return File.is_open(); }
+  truth Eof() { return File.eof(); }
+  void SeekPosBegin(long Offset) { File.seekg(Offset, std::ios::beg); }
+  void SeekPosCurrent(long Offset) { File.seekg(Offset, std::ios::cur); }
+  void SeekPosEnd(long Offset) { File.seekg(Offset, std::ios::end); }
+  long TellPos() { return File.tellg(); }
   ulong TellLine() { return TellLineOfPos(TellPos()); }
   ulong TellLineOfPos(long);
   cfestring& GetFileName() const { return FileName; }
-  void Close() { fclose(Buffer); Buffer = 0; }
+  void Close() { File.close(); }
  private:
   festring ReadNumberIntr (int CallLevel, long *num, truth *isString, truth allowStr, truth PreserveTerminator);
   int HandlePunct(festring&, int, int);
-  FILE* Buffer;
+  std::ifstream File;
   festring FileName;
   const valuemap* ValueMap;
   truth lastWordWasString;

--- a/FeLib/Source/save.cpp
+++ b/FeLib/Source/save.cpp
@@ -35,40 +35,22 @@ outputfile::outputfile(cfestring& FileName, truth AbortOnErr)
     MakePath(DirectoryPath);
   }
 
-  Buffer = fopen(FileName.CStr(), "wb");
+  File.open(FileName.CStr(), std::ios::binary);
 
   if(AbortOnErr && !IsOpen())
     ABORT("Can't open %s for output!", FileName.CStr());
 }
 
-outputfile::~outputfile()
-{
-  if(Buffer)
-    fclose(Buffer);
-}
-
-void outputfile::ReOpen()
-{
-  fclose(Buffer);
-  Buffer = fopen(FileName.CStr(), "ab");
-}
-
 inputfile::inputfile(cfestring& FileName,
                      const valuemap* ValueMap,
                      truth AbortOnErr)
-: Buffer(fopen(FileName.CStr(), "rb")),
+: File(FileName.CStr(), std::ios::binary),
   FileName(FileName),
   ValueMap(ValueMap),
   lastWordWasString(false)
 {
   if(AbortOnErr && !IsOpen())
     ABORT("File %s not found!", FileName.CStr());
-}
-
-inputfile::~inputfile()
-{
-  if(Buffer)
-    fclose(Buffer);
 }
 
 festring inputfile::ReadWord(truth AbortOnEOF)
@@ -88,9 +70,9 @@ int inputfile::HandlePunct(festring& String, int Char, int Mode)
 {
   if(Char == '/')
   {
-    if(!feof(Buffer))
+    if(!Eof())
     {
-      Char = fgetc(Buffer);
+      Char = Get();
 
       if(Char == '*')
       {
@@ -99,11 +81,11 @@ int inputfile::HandlePunct(festring& String, int Char, int Mode)
 
         for(;;)
         {
-          if(feof(Buffer))
+          if(Eof())
             ABORT("Unterminated comment in file %s, beginning at line %ld!",
                   FileName.CStr(), TellLineOfPos(StartPos));
 
-          Char = fgetc(Buffer);
+          Char = Get();
 
           if(OldChar != '*' || Char != '/')
           {
@@ -128,13 +110,13 @@ int inputfile::HandlePunct(festring& String, int Char, int Mode)
       }
       else
       {
-        ungetc(Char, Buffer);
-        clearerr(Buffer);
+        File.putback(Char);
+        File.clear();
       }
     }
 
     if(Mode)
-      ungetc('/', Buffer);
+      File.putback('/');
     else
       String << '/';
 
@@ -143,7 +125,7 @@ int inputfile::HandlePunct(festring& String, int Char, int Mode)
 
   if(Mode)
   {
-    ungetc(Char, Buffer);
+    File.putback(Char);
     return PUNCT_RETURN;
   }
 
@@ -155,11 +137,11 @@ int inputfile::HandlePunct(festring& String, int Char, int Mode)
 
     for(;;)
     {
-      if(feof(Buffer))
+      if(Eof())
         ABORT("Unterminated string in file %s, beginning at line %ld!",
               FileName.CStr(), TellLineOfPos(StartPos));
 
-      Char = fgetc(Buffer);
+      Char = Get();
 
       if(Char != '"')
       {
@@ -186,7 +168,7 @@ void inputfile::ReadWord(festring& String, truth AbortOnEOF)
   String.Empty();
   lastWordWasString = false;
 
-  for(int Char = fgetc(Buffer); !feof(Buffer); Char = fgetc(Buffer))
+  for(int Char = Get(); !Eof(); Char = Get())
   {
     if(isalpha(Char) || Char == '_')
     {
@@ -194,7 +176,7 @@ void inputfile::ReadWord(festring& String, truth AbortOnEOF)
         Mode = MODE_WORD;
       else if(Mode == MODE_NUMBER)
       {
-        ungetc(Char, Buffer);
+        File.putback(Char);
         return;
       }
 
@@ -208,7 +190,7 @@ void inputfile::ReadWord(festring& String, truth AbortOnEOF)
         Mode = MODE_NUMBER;
       else if(Mode == MODE_WORD)
       {
-        ungetc(Char, Buffer);
+        File.putback(Char);
         return;
       }
 
@@ -227,12 +209,12 @@ void inputfile::ReadWord(festring& String, truth AbortOnEOF)
     ABORT("Unexpected end of file %s!", FileName.CStr());
 
   if(Mode)
-    clearerr(Buffer);
+    File.clear();
 }
 
 char inputfile::ReadLetter(truth AbortOnEOF)
 {
-  for(int Char = fgetc(Buffer); !feof(Buffer); Char = fgetc(Buffer))
+  for(int Char = Get(); !Eof(); Char = Get())
   {
     if(isalpha(Char) || isdigit(Char))
       return Char;
@@ -241,9 +223,9 @@ char inputfile::ReadLetter(truth AbortOnEOF)
     {
       if(Char == '/')
       {
-        if(!feof(Buffer))
+        if(!Eof())
         {
-          Char = fgetc(Buffer);
+          Char = Get();
 
           if(Char == '*')
           {
@@ -252,12 +234,12 @@ char inputfile::ReadLetter(truth AbortOnEOF)
 
             for(;;)
             {
-              if(feof(Buffer))
+              if(Eof())
                 ABORT("Unterminated comment in file %s, "
                       "beginning at line %ld!",
                       FileName.CStr(), TellLineOfPos(StartPos));
 
-              Char = fgetc(Buffer);
+              Char = Get();
 
               if(OldChar != '*' || Char != '/')
               {
@@ -281,7 +263,7 @@ char inputfile::ReadLetter(truth AbortOnEOF)
             continue;
           }
           else
-            ungetc(Char, Buffer);
+            File.putback(Char);
         }
 
         return '/';
@@ -336,7 +318,7 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
       if(First == ';' || First == ',' || First == ':')
       {
         if(CallLevel != HIGHEST || PreserveTerminator)
-          ungetc(First, Buffer);
+          File.putback(First);
 
         *num = Value;
         return res;
@@ -345,7 +327,7 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
       if(First == ')')
       {
         if((CallLevel != HIGHEST && CallLevel != 4) || PreserveTerminator)
-          ungetc(')', Buffer);
+          File.putback(')');
 
         *num = Value;
         return res;
@@ -371,7 +353,7 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
               }\
             else\
               {\
-                ungetc(#op[0], Buffer);\
+                File.putback(#op[0]);\
                 *num = Value;\
                 return res;\
               }\
@@ -394,13 +376,13 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
           }
           else
           {
-            ungetc('<', Buffer);
-            ungetc('<', Buffer);
+            File.putback('<');
+            File.putback('<');
             *num = Value;
             return res;
           }
         else
-          ungetc(Next, Buffer);
+          File.putback(Next);
       }
 
       if(First == '>')
@@ -416,20 +398,20 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
           }
           else
           {
-            ungetc('>', Buffer);
-            ungetc('>', Buffer);
+            File.putback('>');
+            File.putback('>');
             *num = Value;
             return res;
           }
         else
-          ungetc(Next, Buffer);
+          File.putback(Next);
       }
 
       if(First == '(')
       {
         if(NumberCorrect)
         {
-          ungetc('(', Buffer);
+          File.putback('(');
           *num = Value;
           return res;
         }
@@ -446,7 +428,7 @@ festring inputfile::ReadNumberIntr(int CallLevel, long *num, truth *isString, tr
 
       if(First == '#') // for #defines
       {
-        ungetc('#', Buffer);
+        File.putback('#');
         *num = Value;
         return res;
       }
@@ -695,7 +677,7 @@ ulong inputfile::TellLineOfPos(long Pos)
   SeekPosBegin(0);
 
   while(TellPos() != Pos)
-    if(fgetc(Buffer) == '\n')
+    if(Get() == '\n')
       ++Line;
 
   if(TellPos() != BackupPos)

